filesystems/memory.py

```diff
@@ -374,8 +375,16 @@ def FS(self, name):
             readlink=_fs(self.readlink),
         )()
 
-    def create_directory(self, path):
-        self[path].create_directory(path=path)
+    def create_directory(self, path, parents=False):
+        if parents:
+            for each in path.heritage():
```

Comment:
I don't think you even need `heritage` here, you should be able to just get away with e.g.:

```
        if parents:
            parent = path.parent()
            try:
                self[parent].create_directory(path=parent, parents=True)
            except exceptions.FileExists:
                pass
        self[path].create_directory(path=path, parents=parents)
```

and shorter even once exists_ok or whatever exists.

(though still a bit of a shame, this API is IMHO an illustrative example of the "boolean flags sux" school of thought, up until now I think the only flag it had was for open, and there I had a plan to kill it with #34, but maybe there's a similar way to eventually kill the flags here too who knows)

Reply:
I expected you to want not-flags.  That said, I can't say I like the idea of creating 572 different function names to cover all the possible permutations of all the flags that might be relevant.  (sure, we're only talking four permutations at the moment but...)

I don't really understand how the traversal can be pushed somewhere else.  What if there exists down to `/a/b` and we want to create `/a/b/c/d/e`.  The parent, `/a/b/c/d` is still a `_NoSuchEntry` and will still be entirely unable to do anything, right?  The filesystem at least needs to 'find' the deepest existing item, if not more.  What you shared did work with the existing test but failed when I made it deeper (just pushed that change).

Reply:
> That said, I can't say I like the idea of creating 572 different function names to cover all the possible permutations of all the flags that might be relevant. (sure, we're only talking four permutations at the moment but...)

Yep, agreed, which is where I think the other kind of API reduces those again (#34) to where you only need 1 per option, not `n * number of options`. But I agree it's easier to just add for now, Filesystems grows fairly slowly considering it has only one library to support at this point.

> What you shared did work with the existing test but failed when I made it deeper (just pushed that change).

Ah! Good! OK, that's another thing I like about this kind of API is that it'll force the right amount of test coverage anyhow -- basically we need one test that hits each kind of file object, so sounds like the first one hit a `_DirectoryChild` but now fails on a `_NoSuchEntry` -- bit of a pain in the ass, but I think basically if you push the idea further, `_NoSuchEntry` which currently does not hold a reference to its parent could grow one (bit of a PITA since currently it's a singleton but bleh) but assuming it grows one, the algorithm is the same as what you have, it's just each object handles its own level -- i.e., `_NoSuchEntry` and `_DirectoryChild` tell their parents to create themselves -- if the parent is a `_NoSuchEntry`, it will continue up the chain.

But possibly you already thought of that and it doesn't work for some reason, so lets see if I can try it myself...

Reply:
It's still not as clean as it could be, but here's at least a quick illustration of what I meant: https://github.com/Julian/Filesystems/commit/501df4cb958b428b533e2b42991377131de8d81c

(But yeah basically we should make sure there's a test that hits each of the different possible leaf objects, which when I write new methods I tend to do by explicitly breaking things by putting `if parents: asdf` and then making sure I've implemented each one).

Reply:
Alrighty, I'm reviewing that and will push soon.  It does feel a bit weird (not specific to this change) that the child creates itself and shoves itself into its parent.  As opposed to the requesting code asking the parent to create a child.  Maybe also the name `.create_as_directory()` rather than the present `.create_directory()` or somesuch would make me feel better.  But this is all just sharing feelings that aren't relevant to this PR.